[FEATURE] Upgrading to DBAL 3 & Switch to symfony/cache

seifane-wise opened the pull request:
### Changes proposed in this pull request:
- Upgrade to DBAL3
- Use symfony/cache for cache since doctrine/cache as been deprecated
- Remove usage from deprecated / deleted classes of DBAL3

I saw too late that a PR was already open concerning this change. I don't know the progress that has been made by @eigan so I opened this one and maybe we can merge our efforts. I think I was able to clear most of the deprecated classes usage in DBAL3. Most significant change is on the cache side where I replaced doctrine/cache by symfony/cache. It broke a lot of tests and I think I managed to fix most of them "correctly". Let me know what you think, if I missed something huge, or if you want to take a different approach on this.

I've been using a version of this branch on my project with no issues so far but I didn't test everything.

Some of the classes I removed (most of them related to codegen) are deprecated on doctrine and will not be replaced by anything. I would argue that if doctrine doesn't want to support codegen anymore it's not the role of this package to implement it. It also keeps the code to a minimum which is easier to in the long run. :) 

Link to the doctrine issue (https://github.com/doctrine/orm/issues/8458)

Link to related PR on laravel-doctrine/migrations (https://github.com/laravel-doctrine/migrations/pull/118)

seifane committed:
Replaced deprecated doctrine/cache
Upgrade to dbal3

seifane committed:
downgrade symfony/cache to 5.4.3 to allow php 7

seifane committed:
bump doctrine/cache version

seifane committed:
bump symfony/cache to 6

seifane committed:
allow php 8.1
fix typo

seifane committed:
Continued cleanup and replacing of deleted classes

eigan commented:
Could you rebase on laravel-doctrine/orm 1.8. Some of your commits (L9) was cherry-picked and thus have a new sha1. The original commits are now trying to be a part of this PR :) 

EDIT: Some of your commits are duplicates of existing commits, for some reason *

eigan commented on src/Configuration/Cache/FileCacheProvider.php:
```diff
@@ -25,13 +25,13 @@ public function __construct(Repository $config)
 
-        return new FilesystemCache(
+        return new FilesystemAdapter(
             $path
```
Third argument is $directory, which I think should be the original $path.

eigan commented on src/Configuration/Cache/PhpFileCacheProvider.php:
```diff
@@ -25,14 +25,15 @@ public function __construct(Repository $config)
-        return new PhpFileCache(
-            $path
+        return new PhpFilesAdapter(
+            $namespace
```
Same with third argument here.

eigan commented on src/Console/ClearMetadataCacheCommand.php:
```diff
@@ -35,20 +34,16 @@ public function handle(ManagerRegistry $registry)
             }
 
-            if ($cache instanceof ApcCache) {
+            if ($cache instanceof IlluminateCacheProvider && $cache->getStore() === "apc") {
```
Could check if instanceof ApcCacheProvider? (there is more of these)

eigan commented:
@dpslwk Any thoughts? I know you tried to use illuminate/cache (which I would prefer, since it is most likely required by all who use this package), but there was some issues?

seifane committed:
Fix style
Correctly pass path to FileCacheProvider and PhpFileCacheProvider
Check apc with instance of instead of getters

seifane-wise commented:
@eigan I rebased from 1.8 but some they were some conflicts on composer.json that I had to edit. I may have messed up something during rebase that cause this weird commit history. I just committed in response to your review.

Concerning the use of symfony/cache it is a suggested dependency of illuminate/cache if that helps my case :). Also it is recommended by the doctrine team and makes integration easier.

seifane committed:
Fix style

eigan commented:
@seifane-wise When you rebase, drop the commits not from you. If using `git rebase -i upstream/1.8` can you replace `pick` with `drop` at the commits. Looks like at least three commits(?) dated march 2.

seifane-wise commented:
@eigan should be good now. commit are correctly rebased.

seifane committed:
Fix doctrine/inflector dependency
Fix test mocking for FileCacheProvider and PhpFileCacheProvider

eigan commented on composer.json:
```diff
@@ -18,9 +18,9 @@
-        "doctrine/orm": "^2.11",
+        "doctrine/cache": "^2",
+        "doctrine/dbal": "^3.3",
+        "doctrine/orm": "~2.11.0",
```
Why change to ~ here?

seifane-wise replied:
No good reason, it's reverted.

eigan commented on composer.json:
```diff
@@ -31,7 +31,9 @@
         "illuminate/view": "^9.0",
-        "symfony/serializer": "^6.0"
+        "symfony/serializer": "^6.0",
+        "doctrine/inflector": "^2",
```
Why was the requirement of inflector ^2 needed?

seifane-wise replied:
It's not needed per se but it was ^1.4 before so I bumped it to ^2

dpslwk replied:
we removed it in the 1.8 branch, (see https://github.com/laravel-doctrine/orm/commit/e97d1454cbfd8a5d14dafe59f9b39a7fe6cb72cf) it was ^1.4 in the other branches, so this has added it back in, not bumped it

seifane-wise replied:
Oh my bad I got confused then :/ @dpslwk do you want me to remove it ?

dpslwk replied:
yeah remove it, its include via `doctrine/orm` so no need to call it out

dpslwk commented:
not sure about the removal of the YAML in this, does that break the yaml mapping for entities? I thought that we would not need to remove that until orm 3.x, not dabl 3.x

seifane committed:
Fix doctrine orm version requirements in composer

seifane-wise commented:
sorry @dpslwk I didn't you had plans to remove it in the future. I had a glance over all deprecated class usage and removed as much as I could as I was trying to future proof as much I could right now. Maybe this branch should also be a new version like my PR on migrations ? 2.x or 3.x ?

SCIF commented:
Hey guys,

First of all, am highly appreciate your work!

It's 2.5 months past last activity. Is there any chance to get this stuff done? Moreover, doctrine team [has advised in January](https://www.doctrine-project.org/2022/01/22/sunsetting-dbal-2.html) to migrate to 3.x branch

chewbakartik commented on src/Configuration/Connections/MasterSlaveConnection.php:
```diff
@@ -2,7 +2,7 @@
 namespace LaravelDoctrine\ORM\Configuration\Connections;
 
-use Doctrine\DBAL\Connections\MasterSlaveConnection as MasterSlaveDoctrineWrapper;
+use Doctrine\DBAL\Connections\PrimaryReadReplicaConnection as MasterSlaveDoctrineWrapper;
```
How challenging would it be to change the name of the alias as well? If I'm understanding the change that Doctrine underwent, `PrimaryReadReplicaConnection` is intending to replace problematic references in programming to slavery. Obviously this isn't something that would be a functional change, but would be in line with many movements that the software field has been addressing with legacy problematic naming conventions.

dpslwk commented:
> sorry @dpslwk I didn't you had plans to remove it in the future. I had a glance over all deprecated class usage and removed as much as I could as I was trying to future proof as much I could right now. Maybe this branch should also be a new version like my PR on migrations ? 2.x or 3.x ?

I really think the YAML meta data driver should be restored, that is deprecated with ORM and will be removed in ORM 3.x its not related to DBAL 3
Yes we will need to address this when ORM 3 is eventually released
Side note my main project that uses ORM and the main reason for my work here currently uses YAML mappings, the longer I can put off having to rewrite 63 mappings the better ;)

seifane-wise commented:
Sorry I am closing this PR as I dont have the time to maintain it anymore. I will leave my branch open if someone want to base off of it. :pray:

eigan commented:
@seifane-wise Understandable! Thanks! 🙂
